go-controller/hybrid-overlay/cmd/hybrid-overlay-node/hybrid-overlay-node.go

```diff
@@ -38,40 +39,52 @@ func main() {
 			Destination: &nodeName,
 		},
 		&cli.BoolFlag{
-			Name:        "windows-service",
-			Usage:       "Enables hybrid overlay to run as a Windows service. Ignored on Linux.",
-			Destination: &runAsWindowsService,
+			Name:  "windows-service",
+			Usage: "Enables hybrid overlay to run as a Windows service. Ignored on Linux.",
 		}})
+
+	for _, flag := range c.Flags {
+		if strings.Contains(flag.String(), "windows-service") {
+			runAsWindowsService = true
+		}
+	}
+
+	ctx := context.Background()
+
 	c.Action = func(c *cli.Context) error {
 		if err := runHybridOverlay(c); err != nil {
 			panic(err.Error())
 		}
 		return nil
 	}
 
-	ctx := context.Background()
-
-	// trap SIGHUP, SIGINT, SIGTERM, SIGQUIT and
-	// cancel the context
-	ctx, cancel := context.WithCancel(ctx)
-	exitCh := make(chan os.Signal, 1)
-	signal.Notify(exitCh,
-		syscall.SIGHUP,
-		syscall.SIGINT,
-		syscall.SIGTERM,
-		syscall.SIGQUIT)
-	defer func() {
-		signal.Stop(exitCh)
-		cancel()
-	}()
-	go func() {
-		select {
-		case s := <-exitCh:
-			klog.Infof("Received signal %s. Shutting down", s)
-			cancel()
-		case <-ctx.Done():
+	if runAsWindowsService {
+		if err := initForOS(&ctx); err != nil {
+			klog.Exit(err)
 		}
-	}()
+	} else {
+		// trap SIGHUP, SIGINT, SIGTERM, SIGQUIT and
+		// cancel the context
+		ctx, cancel := context.WithCancel(ctx)
```

Comment:
ctx is defined as a new value here. Does cancel() work properly in this case?

Reply:
Clarifying my comment here... The value we redefine at this location is different from the value we pass to RunContext at line 89 below. I know WithCancel() returns a cancellable copy of the context that was passed as a parameter (i.e. the parent being the context created at line 52 above), but I don't know if cancelling the value defined at this location acts on the parent value defined at line 52. Unfortunately, I couldn't find any documentation on this topic. Please do try it out and make sure there is no regression.

Reply:
Thanks for your review @JocelynBerrendonner , appreciate it! Cancel works properly in this case and it cancels the parent context as well when the child context is cancelled. I have tested it and did not run into any issues. I was also planning to add some restructuring changes which could remove the unnecessary contexts as well. Perhaps I can add them as another commit and you can take a look at them.

Reply:
Sounds good! Thanks for confirming, @mansikulkarni96 ! I am good either way, feel free to take the approach that is the easiest for you!

Reply:
@JocelynBerrendonner Thank you! I have moved the context cancellation declaration-> `ctx, cancel := context.WithCancel(ctx)`  outside of if else, as it was the easiest way to ensure the same context is passed to RunContext, the code restructuring seems complex as we don't have much control over the Execute method so I am going to leave it as it is.

Reply:
np! Thank you @mansikulkarni96 !